src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspace.cs

```diff
@@ -114,5 +114,16 @@ public PortableExecutableReference CreatePortableExecutableReference(string file
         }
 
         internal abstract string TryGetRuleSetPathForProject(ProjectId projectId);
+
+        /// <summary>
+        /// Attempts to sync a named type name and namespace based on file path. PreviousFilePath is 
+        /// used to determine how the policy might be applied, assuming that matching types with previous
+        /// indicates a user wants the type to match with the new location or name.
+        /// May fail if:
+        /// - The appropriate type for naming policy cannot be found
+        /// - The type is partial and spread across multiple files
+        /// - The solution changes during the operation
+        /// </summary>
+        public abstract void TrySyncNamedTypeToFilePath(DocumentId documentId, string previousFilePath);
```

Comment:
What does failure mean here? An exception?

Also, why wouldn't a rename of a partial type succeed?

Reply:
How does this happen: The solution changes during the operation
?

Reply:
Failure here shouldn't be an exception, do we need to track if failure happens? 

> Solution changes during the operation
This could potentially kick off async work such as user dialog to confirm intent. We don't want to block on this call as changes are applied, afaik. There is a requirement that the intent of this API is for project-system work, and blocking is a nonstarter for them. Would addition of optional cancellation token help convey that?

Reply:
>The solution changes during the operation
yes a `git checkout` for example

Reply:
>Would addition of optional cancellation token help convey that?

I would prefer that roslyn does everything eventually. It knows much more about rename than we do. If we own the threaded wait dialog that the user clicks on to cancel rename (which I don't want to own) then we should have a `CancelationToken` otherwise its fine as is.

I just wanted to make sure that the intent was that this API would not throw so the caller does not need to consider this case.

Reply:
`public abstract void TrySyncNamedTypeToFilePath(DocumentId documentId, string previousFilePath, CancellationToken cancellationToken = CancellationToken.None);` 

That seems reasonable to convey that something is long enough and could be cancelled, but you opt into caring.

Reply:
> yes a git checkout for example

Gotcha.  Note that the workspace at least provides information about this in a return parameter when this happens.  Should we not do the same here?

> There is a requirement that the intent of this API is for project-system work, and blocking is a nonstarter for them

If this is intended to be non-blockin (and it takes a CT), should this not just be an async method with a CT?  Thanks!

Reply:
`async method`

hmm yeah it should either be async or non-cancelable. If its not async then I expect it to return immediately. I don't really want to burn a background thread waiting for this to complete

Reply:
plus, it tickles me to have a method called `TrySyncXXXAsync` :D

Maybe `TryUpdateSymbolsToMatchFileNameAsync`?

Reply:
>Gotcha. Note that the workspace at least provides information about this in a return parameter when this happens. Should we not do the same here?

As I am visualizing consuming this on the project system side, I do think I would want to have the method return whether the rename applied or failed so I will know if I need to notify the user of anything

Reply:
Do you need to distinguish between "this failed due to an actual problem" and "there just wasn't a symbol to rename?"

Reply:
Actual problem I assume would be an exception, but maybe we don't do that for our public APIs? 

Otherwise, from our discussions I don't think differentiating would matter.